### PR TITLE
Fix logistic regression model

### PR DESCRIPTION
**Pull request by TomasVaskevicius**

### CODE REVIEW

File: firefly/DESCRIPTION [hunk 1]
```diff
@@ -2,8 +2,8 @@ Package: firefly
 Title: FireFly Monte Carlo implementation
 Version: 0.0.0.9000
 Authors@R: c(
-  person("Francesca", "Crucinio"),
-  person("Tomas", "Vaskevicius"))
+  person("Francesca", "Crucinio", email = "francesca.crucinio@spc.ox.ac.uk", role = c("aut")),
```
Comment (TomasVaskevicius) [inline firefly/DESCRIPTION]: I had to add our emails, otherwise devtools wouldn't work.

File: firefly/R/logistic_regression.R [hunk 2]
```diff
@@ -1,99 +1,107 @@
...
-  LogBound <- function(x, theta, map.tuned, t){
-    N <- ncol(x)
-    if(is.null(map.tuned) {
-      xi <- rep(1.5,N)
-    }
-    else{
-    xi <- t(map.tuned)%*%(x*t)
-    }
-   
-    a <- -(exp(xi)-1)/(4*xi*(exp(xi)+1))
```
Comment (TomasVaskevicius) [inline firefly/R/logistic_regression.R]: I think there was a mistake in the paper when computing the a. 
Check out the https://pdfs.semanticscholar.org/e407/ea7fda6d152d2186f4b5e27aa04ec2d32dcd.pdf.
My change is consistent with how they compute "a" in their python code I think.

File: firefly/R/logistic_regression.R [hunk 3]
```diff
@@ -1,99 +1,107 @@
...
-    log.likelihood <- log(inv.logit((t(theta)%*%x)*t))
-    return(log.likelihood)
+
+  # Define a defaul vector xi, specifying where each log likelihood lower bound
+  # should be tight.
+  xi <- rep(1.5, nrow(X))
+  if (theta.map.estimate != NULL) {
+    # We can come up with a better value of xi.
+    # Implement this later.
+    stop(paste("Using theta.map.estimate int logistic regression model",
```
Comment (TomasVaskevicius) [inline firefly/R/logistic_regression.R]: Let's only worry about tuned parameters once we have something working.

File: firefly/R/logistic_regression.R [hunk 4]
```diff
@@ -1,99 +1,107 @@
...
+  LogLikelihood <- function(ids, theta){
+    # Computes the log likelihood for a subset of our data.
+    #
+    # @param ids A vector of data point ids, specifying the subset of our
+    #            dataset.
+    # @param theta Model parameters, where the log likelihoods should be
+    #        evaluated.
+    # @return A vector of log likelihoods for the requested subset of our
+    #         dataset.
+    log(1 / (1 + exp((X[ids,] %*% theta) * y[ids])))
```
Comment (FrancescaCrucinio) [inline firefly/R/logistic_regression.R]: There's a - missing 
built in function inv.logit would work faster

File: firefly/R/logistic_regression.R [hunk 5]
```diff
@@ -1,99 +1,107 @@
...
+    # @return A vector of log bounds for the requested subset of our
+    #         dataset.
+    X.theta.y <- (X[ids,] %*% theta) * y[ids]
+    a[ids] * (X.theta.y)^2 + b[ids] * X.theta.y + c[ids]
   }
-  
-  ##Compute log pseudo prior
-  LogPseudoPrior <- function(x, theta, map.tuned, scale, t){ 
-    log.prior <- sum(dnorm(theta, mu, scale, log=TRUE)) #compute log prior
-    log.bound <- LogBound(x, theta, map.tuned, t)
```
Comment (TomasVaskevicius) [inline firefly/R/logistic_regression.R]: This is too slow, there is a more effiient way to compute the log product of bounds for all the dataset! This is precisely why the bound functions have to be chosen in a way to make the computation of their product over all of the dataset quick.

File: firefly/R/logistic_regression.R [hunk 6]
```diff
@@ -1,99 +1,107 @@
...
+    #            dataset.
+    # @param theta Model parameters, where the log differences should be
+    #        evaluated.
+    # @return A vector of log differences for the requested subset of our
+    #         dataset.
+    likelihoods <- 1 / (1 + exp((X[ids,] %*% theta) * y[ids]))
+    log(likelihoods - exp(LogBound(ids, theta)))
   }
-  
-  LogGradient <-function(x, theta, map.tuned, z, t){
```
Comment (TomasVaskevicius) [inline firefly/R/logistic_regression.R]: I have removed this for a while. Looks way too complicated and we first need to focus on getting something to work. After that we will put this function back and test it.

File: firefly/R/logistic_regression.R [hunk 7]
```diff
@@ -1,99 +1,107 @@
...
-  
-  ##Compute log likelihood
-  LogLikelihood <- function(x, theta, t){
-    log.likelihood <- log(inv.logit((t(theta)%*%x)*t))
-    return(log.likelihood)
+
+  # Define a defaul vector xi, specifying where each log likelihood lower bound
+  # should be tight.
+  xi <- rep(1.5, nrow(X))
+  if (theta.map.estimate != NULL) {
```
Comment (FrancescaCrucinio) [inline firefly/R/logistic_regression.R]: Robin Evans suggested to use is.null() instead of !=

File: firefly/R/logistic_regression.R (hunk 4 above)
Comment (TomasVaskevicius) [reply firefly/R/logistic_regression.R]: You are right! Will fix in a minute.

File: firefly/R/logistic_regression.R [hunk 8]
```diff
@@ -1,99 +1,107 @@
...
+  # We use the lower bound due to Jaakkola and Jordan (1996)
+  # "A variational approach to Bayesian logistic regression models and
+  # their extensions".
+  # Note that we compute a different bound for each of the data point.
+  sig.xi <- 1 / (1 + exp(-xi))
+  sig.neg.xi <- 1 / (1 + exp(xi))
+  a <- (sig.neg.xi - sig.xi) / (4 * xi)
+  b <- 1/2
+  c <- -a * xi^2 - xi / 2 + log(sig.xi)
+
```
Comment (FrancescaCrucinio) [inline firefly/R/logistic_regression.R]: There are 2 version for computing a =O
paper and phd thesis give the same value (the one i used), but Python code gives another one
no idea of which is correct

File: firefly/R/logistic_regression.R [hunk 9]
```diff
@@ -1,99 +1,107 @@
...
+    # Computes the log of likelihood and lower bound difference for a given
+    # subset of our dataset.
+    #
+    # @param ids A vector of data point ids, specifying the subset of our
+    #            dataset.
+    # @param theta Model parameters, where the log differences should be
+    #        evaluated.
+    # @return A vector of log differences for the requested subset of our
+    #         dataset.
+    likelihoods <- 1 / (1 + exp((X[ids,] %*% theta) * y[ids]))
```
Comment (FrancescaCrucinio) [inline firefly/R/logistic_regression.R]: see above (- missing)

File: firefly/R/logistic_regression.R [hunk 10]
```diff
@@ -1,99 +1,107 @@
...
+    # Computes the log lower bound for likelihoods of a subset of our data.
+    #
+    # @param ids A vector of data point ids, specifying the subset of our
+    #            dataset.
+    # @param theta Model parameters, where the log bounds should be
+    #        evaluated.
+    # @return A vector of log bounds for the requested subset of our
+    #         dataset.
+    X.theta.y <- (X[ids,] %*% theta) * y[ids]
+    a[ids] * (X.theta.y)^2 + b[ids] * X.theta.y + c[ids]
```
Comment (FrancescaCrucinio) [inline firefly/R/logistic_regression.R]: I don't think the indexes in b are needed, +b should work fine

File: firefly/R/logistic_regression.R [hunk 11]
```diff
@@ -1,99 +1,107 @@
...
+  LogLikelihood <- function(ids, theta){
+    # Computes the log likelihood for a subset of our data.
+    #
+    # @param ids A vector of data point ids, specifying the subset of our
+    #            dataset.
+    # @param theta Model parameters, where the log likelihoods should be
+    #        evaluated.
+    # @return A vector of log likelihoods for the requested subset of our
+    #         dataset.
+    log(plogis((X[ids,] %*% theta) * y[ids]))
```
Comment (FrancescaCrucinio) [inline firefly/R/logistic_regression.R]: is this invers logit? I think is just logit

**Comment by TomasVaskevicius** [comment]
Hi Francesca,

I have fixed the likelihood computation to use the native code (inv.logit function that you suggested just calls plogis, so this is what I do now).

Also changed the != NULL like to !is.null()

**Pull request merged**